### PR TITLE
Make RegExp.prototype.toString generic

### LINKED ISSUE #578: RegExp.prototype.toString should be generic
Rather than requiring `this` to be a RegExp object, the `toString` method should be generic and use the `source` and `flags` properties to generate the string. Kangax test case:

```JS
// RegExp.prototype.toString -> Get -> [[Get]]
var get = [];
var p = new Proxy({}, { get: function(o, k) { get.push(k); return o[k]; }});
RegExp.prototype.toString.call(p);
return get + '' === "source,flags";
```

### PR DESCRIPTION
**Pull request by goyakin**
Fixes #578

### CODE REVIEW

**Comment by goyakin** [comment]
@akroshg & @abchatra, could you take a look?

**Comment by goyakin** [comment]
Benchmarks show no difference in perf.

**Comment by goyakin** [comment]
Ping...

File: lib/Runtime/Library/JavascriptRegularExpression.cpp [hunk 1]
```diff
@@ -671,11 +671,37 @@ namespace Js
...
+            builder->Append(flags);
+            return builder;
+        }
+        else
+        {
+            JavascriptRegExp* obj = GetJavascriptRegExp(args, varName, scriptContext);
+
+            bool sourceOnly = false;
+            bool useFlagsProperty = scriptConfig->IsES6RegExPrototypePropertiesEnabled();
+            return obj->ToString(sourceOnly, useFlagsProperty);
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptRegularExpression.cpp]: Is this the only place the ::ToString gets called? wondering if we can remove useFlagProperty altogether. In the ToString itself can check of IsES6RegExPrototypePropertiesEnabled() and do the work.

File: test/es6/regex-prototype-properties.js [hunk 2]
```diff
@@ -79,5 +79,20 @@ tests = tests.concat(nonGenericPropertyNames.map(function (name) {
...
+tests = tests.concat([
+    {
+        name: "RegExp.prototype.toString should be generic",
+        body: function () {
+            var re = {
+                source: "source",
+                flags: "gi"
+            };
+
+            var string = RegExp.prototype.toString.call(re);
```
Comment (akroshg) [inline test/es6/regex-prototype-properties.js]: What if the source and flags are not there, say RegExp.prototype.toString.call({}) ?

**Comment by akroshg** [comment]
code wise looks good to me.

File: lib/Runtime/Library/JavascriptRegularExpression.cpp (hunk 1 above)
Comment (goyakin) [inline lib/Runtime/Library/JavascriptRegularExpression.cpp]: This isn't the only place. It is also used to generate the string for the `source` property. However, now that `EntryToString` handles the `flags` property itself, the part of `ToString` handling it is unreachable. I've removed that section in commit 154d829011cd3290c7ba15a60bad3df250d5936b.

File: test/es6/regex-prototype-properties.js (hunk 2 above)
Comment (goyakin) [inline test/es6/regex-prototype-properties.js]: We should use the string `undefined` for them in that case. I didn't add a test specifically for this, but added some verifying the function when the properties are `undefined` in addition to non-string properties and when they are overridden on a subclass (ee246cce09c70b33b195c631050419b2785d8874).

**Comment by akroshg** [comment]
Looks good to me.

**Pull request merged**